daemon/volumes_unix.go

```diff
@@ -52,11 +53,13 @@ func (daemon *Daemon) setupMounts(c *container.Container) ([]container.Mount, er
 			return nil, err
 		}
 		if !c.TrySetNetworkMount(m.Destination, path) {
+			nonRecursive := m.Spec.Type == mounttypes.TypeBind && m.Spec.BindOptions != nil && m.Spec.BindOptions.NonRecursive
```

Comment:
This is a bit hard to read; perhaps;

```suggestion
			var nonRecursive bool
			if m.Spec.Type == mounttypes.TypeBind && m.Spec.BindOptions != nil {
				nonRecursive = m.Spec.BindOptions.NonRecursive
			}
```

Or perhaps even better; move it lower, then the intermediate variable is not even needed;

```suggestion
			mnt := container.Mount{
				Source:      path,
				Destination: m.Destination,
				Writable:    m.RW,
				Propagation: string(m.Propagation),
				Source:       path,
				Destination:  m.Destination,
				Writable:     m.RW,
				Propagation:  string(m.Propagation),
			}
			if m.Spec.Type == mounttypes.TypeBind && m.Spec.BindOptions != nil {
				mnt.NonRecursive = m.Spec.BindOptions.NonRecursive
			}
```

Reply:
Looking at the above; `NonRecursive = true` is only supported for `type=bind`, we should add that `ValidateMountConfig()`;

https://github.com/moby/moby/blob/8e610b2b55bfd1bfa9436ab110d311f5e8a74dcb/volume/mounts/linux_parser.go#L50-L110

and LCOW / Windows;

https://github.com/moby/moby/blob/8e610b2b55bfd1bfa9436ab110d311f5e8a74dcb/volume/mounts/lcow_parser.go#L10-L26

https://github.com/moby/moby/blob/8e610b2b55bfd1bfa9436ab110d311f5e8a74dcb/volume/mounts/windows_parser.go#L216-L299